lib/collection/src/tests/shard_query.rs

```diff
@@ -0,0 +1,147 @@
+        CollectionError::bad_request("cannot make RRF without prefetches".to_string());
+    assert!(matches!(sources_scores, Err(err) if err == expected_error));
+
+    // RFF query with prefetches
+    let nearest_query = QueryEnum::Nearest(NamedVectorStruct::new_from_vector(
+        Vector::Dense(vec![1.0, 2.0, 3.0, 4.0]),
+        DEFAULT_VECTOR_NAME,
+    ));
+    let inner_limit = 3;
+    let nearest_query_prefetch = ShardPrefetch {
+        prefetches: vec![], // no recursion here
+        query: ScoringQuery::Vector(nearest_query.clone()),
+        limit: inner_limit,
+        params: None,
+        filter: None,
+        score_threshold: None,
+    };
+    let outer_limit = 2;
+    let query = ShardQueryRequest {
+        prefetches: vec![nearest_query_prefetch.clone()],
+        query: ScoringQuery::Rrf,
+        filter: None,
+        score_threshold: None,
+        limit: outer_limit,
+        offset: 0,
+        params: None,
+        with_vector: WithVector::Bool(false),
+        with_payload: WithPayloadInterface::Bool(false),
+    };
+
+    let sources_scores = shard
+        .query(Arc::new(query), &current_runtime)
+        .await
+        .unwrap();
+
+    // only one inner prefetch
+    assert_eq!(sources_scores.len(), 1);
+    // number of results is limited by the outer limit for rescoring
+    assert_eq!(sources_scores[0].len(), outer_limit);
+
+    // rescoring against a vector without prefetches
+    let outer_limit = 2;
+    let query = ShardQueryRequest {
+        prefetches: vec![],
+        query: ScoringQuery::Vector(nearest_query.clone()),
+        filter: None,
+        score_threshold: None,
+        limit: outer_limit,
+        offset: 0,
+        params: None,
+        with_vector: WithVector::Bool(false),
+        with_payload: WithPayloadInterface::Bool(false),
+    };
+
+    let sources_scores = shard
+        .query(Arc::new(query), &current_runtime)
+        .await
+        .unwrap();
+
+    // only one inner result in absence of prefetches
+    assert_eq!(sources_scores.len(), 1);
+    // number of results is limited by the outer limit for rescoring
+    assert_eq!(sources_scores[0].len(), outer_limit);
+
+    // rescoring against a vector with prefetches
+    let outer_limit = 2;
+    let query = ShardQueryRequest {
+        prefetches: vec![nearest_query_prefetch],
+        query: ScoringQuery::Vector(nearest_query),
+        filter: None,
+        score_threshold: None,
+        limit: outer_limit,
+        offset: 0,
+        params: None,
+        with_vector: WithVector::Bool(false),
+        with_payload: WithPayloadInterface::Bool(false),
+    };
+
+    let sources_scores = shard
+        .query(Arc::new(query), &current_runtime)
+        .await
+        .unwrap();
+
+    // only one inner result in absence of prefetches
+    assert_eq!(sources_scores.len(), 1);
+    // number of results is limited by the outer limit for rescoring
+    assert_eq!(sources_scores[0].len(), outer_limit);
+}
```

Comment:
I would love to see a the case of 2+ prefetches with RRF rescore, to show that the output has two lists, not just one.